lib/archethic/self_repair/notifier.ex

```diff
@@ -1,173 +1,72 @@
-        state
-      )
-      when date != nil do
-    {:noreply, Map.update!(state, :notified, &Map.delete(&1, node_key))}
-  end
-
-  def handle_info(_, state) do
-    {:noreply, state}
-  end
-
-  defp repair_transactions(node_key, current_node_public_key) do
-    Logger.debug("Trying to repair transactions due to a topology change",
-      node: Base.encode16(node_key)
-    )
-
-    node_key
-    |> get_transactions_to_sync()
-    |> Stream.each(&forward_transaction(&1, current_node_public_key))
-    |> Stream.run()
-  end
-
-  defp get_transactions_to_sync(node_public_key) do
-    # We fetch all the transactions existing and check if the disconnecting node was a storage node
-    TransactionChain.list_all([:address, :type, validation_stamp: [:timestamp]])
-    |> Stream.map(
-      fn tx = %Transaction{
-           address: address,
-           type: type,
-           validation_stamp: %ValidationStamp{timestamp: timestamp}
-         } ->
-        node_list =
-          Enum.filter(
-            P2P.list_nodes(),
-            &(&1.authorization_date != nil and
-                DateTime.compare(&1.authorization_date, timestamp) == :lt)
-          )
-
-        {tx, Election.chain_storage_nodes_with_type(address, type, node_list)}
-      end
-    )
-    |> Stream.filter(fn {_tx, nodes} ->
-      Utils.key_in_node_list?(nodes, node_public_key)
-    end)
+  @doc """
+  Start a new RepairWorker for the first_address
+  """
+  @spec start_worker(ShardRepair.t()) :: DynamicSupervisor.on_start_child()
+  def start_worker(msg) do
+    DynamicSupervisor.start_child(NotifierSupervisor, {RepairWorker, msg})
   end
 
-  defp forward_transaction(
-         {tx = %Transaction{address: address, type: type}, previous_storage_nodes},
-         current_node_public_key
-       ) do
-    # We compute the new storage nodes minus the previous ones
-    new_storage_nodes =
-      Election.chain_storage_nodes_with_type(
-        address,
-        type,
-        P2P.authorized_nodes() -- previous_storage_nodes
-      )
-
-    with false <- Enum.empty?(new_storage_nodes),
-         true <- Utils.key_in_node_list?(previous_storage_nodes, current_node_public_key) do
-      Logger.info("Repair started due to network topology change",
-        transaction_address: Base.encode16(address),
-        transaction_type: type
-      )
-
-      Task.Supervisor.async_stream_nolink(
-        TaskSupervisor,
-        new_storage_nodes,
-        &P2P.send_message(&1, %ReplicateTransaction{transaction: tx}),
-        ordered: false,
-        on_timeout: :kill_task
-      )
-      |> Stream.run()
-    end
+  @doc """
+  Add a new address in the address list of the RepairWorker
+  """
+  @spec add_message(pid(), ShardRepair.t()) :: :ok
+  def add_message(pid, %ShardRepair{
+        storage_address: storage_address,
+        io_addresses: io_addresses
+      }) do
+    GenServer.cast(pid, {:add_address, storage_address, io_addresses})
```

Comment:
We should use abstraction with the module name instead of giving implementation details.